tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs

```diff
@@ -26,18 +27,31 @@ namespace MongoDB.Bson.Tests.Jira
-        [InlineData(typeof(DerivedWithoutSetterAndWithBsonElementAndWithPrivateConstructor_BaseWithoutSetter))]
-        [InlineData(typeof(DerivedWithoutSetterAndWithoutBsonElement_AbstractBaseWithoutSetterAndWithProtectedConstructor))]
-        [InlineData(typeof(DerivedWithoutSetterAndWithoutBsonElement_AbstractBaseWithoutSetter))]
-        [InlineData(typeof(DerivedWithPrivateSetterAndWithBsonElement_BaseWithoutSetter))]
-        public void Serialization_should_work_as_expected(Type testCaseType)
+        [InlineData(typeof(DerivedWithoutSetter_BaseWithoutSetter), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedWithoutSetter_BaseWithPrivateSetterAndWithProtectedConstructor), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedWithoutSetterAndWithBsonElement_BaseWithoutSetterAndWithProtectedConstructor), new[] { 1, 2 }, "{ \"X\" : 1, \"y\" : 2 }")]
+        [InlineData(typeof(DerivedWithoutSetterAndWithBsonElementAndWithPrivateConstructor_BaseWithoutSetter), new[] { 1, 2 }, "{ \"X\" : 1, \"y\" : 2 }")]
+        [InlineData(typeof(DerivedWithoutSetterAndWithoutBsonElement_AbstractBaseWithoutSetterAndWithProtectedConstructor), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedWithoutSetterAndWithoutBsonElement_AbstractBaseWithoutSetter), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedWithPrivateSetterAndWithBsonElement_BaseWithoutSetter), new[] { 1, 2 }, "{ \"X\" : 1, \"y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorAndWithBsonConstructorAttribute_AbstractBaseImmutable), new[] { 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorAndWithBsonConstructorAttribute_AbstractBaseImmutableWithConstructor), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2  }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonElementAttributeAndWithBsonConstructorAttribute_AbstractBaseImmutable), new[] { 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonElementAttributeAndWithBsonConstructorAttribute_AbstractBaseImmutableWithConstructor), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonIgnoreAttribute_AbstractBaseImmutableWithAbstractProperty), new[] { 2 }, "{ \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonIgnoreAttribute_AbstractBaseImmutableWithConstructor), new[] { 1 }, "{ \"X\" : 1 }")]
+        [InlineData(typeof(ClassWithTwoConstructorsWhereTheFirstIsNotFullyMatchedAndTheSecondIsFullyMatched), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(ClassWithOneConstructorThatIsNotFullyMatched), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2, \"Z\" : null }")]
```

Comment:
This comment was lost when you renamed the file:

It's worth asking whether Z should be serialized or not.

We might want to make a small change to ImmutableTypeClassMapConvention (line 88) so that only properties that can be matched to a constructor argument (any constructor) will be mapped.

This used to not be an issue because we used to only consider the one constructor (if any) that had the same number of arguments as the class had properties.

Reply:
>It's worth asking whether Z should be serialized or not.

I think it should be serialized. Because the class can be written like this:

        public class ClassWithOneConstructorThatIsNotFullyMatched
        {
            public int? X { get; }
            public int? Y { get; }
            public int? Z { get; } = 2;  // The default value

            public ClassWithOneConstructorThatIsNotFullyMatched(int? x, int? y)
            {
                X = x;
                Y = y;
            }
        }

and in this case, we will have this serialized value: `{X=1, Y=2, Z=2}.`

Reply:
The counter argument is that it probably won't round trip because `Z` in the document can't be matched to any constructor argument.

Reply:
When reading the document the Z value will be ignored (even though I expected an exception to be thrown).

So the following executes without an exception:

    var json = "{ X : 1, Y : 2, Z : 123456 }"
    var result = BsonSerializer.Deserialize(json, testCaseType);

BUT... the Z value of the result will be `2` (or whatever the class sets it to).

I don't think we want to write values OUT to the database that we can't read back IN.

Reply:
I'm willing to ignore this issue for now if you and @JamesKovacs both argue in favor of ignoring it.

Reply:
Maybe we need to allow saving all fields and not only mentioned in the constructor then? But it will require additional investigation, I can create a ticket to do it

Reply:
One of our goals in serialization is to no silently lose data.

To achieve that goal we should NOT include `Z` in the serialized form because it is ONE WAY.

I don't think a new ticket is needed... but would like to hear @JamesKovacs  opinion as well.

Reply:
Note that we already have precedent (in mutable classes) for NOT serializing properties that can't be read back in.